extensions/api/data-management/policydefinition/src/test/java/org/eclipse/dataspaceconnector/api/datamanagement/policy/service/PolicyServiceImplTest.java

```diff
@@ -0,0 +1,136 @@
+import org.eclipse.dataspaceconnector.spi.policy.store.PolicyStore;
+import org.eclipse.dataspaceconnector.spi.query.QuerySpec;
+import org.eclipse.dataspaceconnector.spi.transaction.NoopTransactionContext;
+import org.eclipse.dataspaceconnector.spi.transaction.TransactionContext;
+import org.eclipse.dataspaceconnector.spi.types.domain.contract.offer.ContractDefinition;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.eclipse.dataspaceconnector.api.result.ServiceFailure.Reason.CONFLICT;
+import static org.eclipse.dataspaceconnector.api.result.ServiceFailure.Reason.NOT_FOUND;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+public class PolicyServiceImplTest {
+
+    private final PolicyStore policyStore = mock(PolicyStore.class);
+    private final ContractDefinitionStore contractDefinitionStore = mock(ContractDefinitionStore.class);
+    private final TransactionContext dummyTransactionContext = new NoopTransactionContext();
+
+    private final PolicyServiceImpl policyService = new PolicyServiceImpl(dummyTransactionContext, policyStore, contractDefinitionStore);
+
+    @Test
+    void findById_shouldRelyOnPolicyStore() {
+        when(policyStore.findById("policyId")).thenReturn(createPolicy("policyId"));
+
+        var policy = policyService.findById("policyId");
+        var uidTest = policy.getUid();
+
+        String assetId = "policyId";
+        assertThat(policy).isNotNull().matches(hasId(assetId));
+    }
+
+    @Test
+    void query_shouldRelyOnPolicyStore() {
+        var policy = createPolicy("policyId");
+        when(policyStore.findAll(any(QuerySpec.class))).thenReturn(Stream.of(policy));
+        var policies = policyService.query(QuerySpec.none());
+
+        assertThat(policies).hasSize(1).first().matches(hasId("policyId"));
+    }
+
+    @Test
+    void createPolicy_shouldCreatePolicyIfItDoesNotAlreadyExist() {
+        var policy = createPolicy("policyId");
+        when(policyStore.findById("policyId")).thenReturn(null);
+
+        var inserted = policyService.create(policy);
+
+        assertThat(inserted.succeeded()).isTrue();
+        assertThat(inserted.getContent()).matches(hasId("policyId"));
+    }
+
+    @Test
+    void createPolicy_shouldNotCreatePolicyIfItAlreadyExists() {
+        var policy = createPolicy("policyId");
+        when(policyStore.findById("policyId")).thenReturn(policy);
+
+        var inserted = policyService.create(policy);
+
+        assertThat(inserted.succeeded()).isFalse();
+    }
+
+    @Test
+    void delete_shouldDeletePolicyIfItsNotReferencedByAnyContractDefinition() {
+        when(contractDefinitionStore.findAll(any())).thenReturn(Stream.empty()).thenReturn(Stream.empty());
```

Comment:
nit: second call to `thenReturn` is unnnecessary. Subsequent calls will keep returning the value from the last `thenReturn`

Reply:
@algattik the problem is that the Stream can be consumed once, and the `thenReturn` returns the same instance, so removing the second `thenReturn` will throw exception

Reply:
@algattik without the second thenReturn I received the exception "java.lang.IllegalStateException: stream has already been operated upon or closed" 

I can do however .thenReturn(Stream.empty(), Stream.empty()) which also works and only uses one thenReturn

Reply:
Interesting :)